.github/workflows/build_and_test.yml

```diff
@@ -1,44 +1,82 @@
-    steps:
-      - name: Checkout
-        uses: actions/checkout@v3
-      - name: Set up Docker Buildx
-        uses: docker/setup-buildx-action@v2
-      - name: Build and export
-        uses: docker/build-push-action@v3
-        with:
-          context: .
-          file: ./Dockerfile
-          tags: meals-api:latest
-          outputs: type=docker,dest=/tmp/image.tar
-
-      - name: Write build result to file
-        if: always()
-        run: |
-          datetime=$(date +"%Y-%m-%d %H:%M:%S")
-          names="Daniel, Omer, Shachar"
-          echo $datetime >> build_result.txt
-          echo $names >> build_result.txt
-          if [ -f /tmp/image.tar ]; then
-            echo "Image successfully built" >> build_result.txt
-          else
-            echo "Image not able to be built" >> build_result.txt
-          fi
-
-      - name: Upload image artifact
-        uses: actions/upload-artifact@v3
-        with:
-          name: meals-api
-          path: /tmp/image.tar
-
-      - name: Upload result file
-        if: always()
-        uses: actions/upload-artifact@v3
-        with:
-          name: meals-api-build-result.txt
-          path: build_result.txt
\ No newline at end of file
+name: build-and-test
+on: [push]
+
+jobs:
+  build:
+    runs-on: ubuntu-latest
+    steps:
+      - name: Checkout
+        uses: actions/checkout@v3
+      - name: Set up Docker Buildx
+        uses: docker/setup-buildx-action@v2
+      - name: Build and export
+        uses: docker/build-push-action@v3
+        with:
+          context: .
+          file: ./Dockerfile
+          tags: meals-api:latest
+          outputs: type=docker,dest=/tmp/image.tar
+
+      - name: Write build result to file
+        if: always()
+        run: |
+          datetime=$(date +"%Y-%m-%d %H:%M:%S")
+          names="Daniel, Omer, Shachar"
+          echo $datetime >> build_result.txt
+          echo $names >> build_result.txt
+          if [ -f /tmp/image.tar ]; then
+            echo "Image successfully built" >> build_result.txt
+          else
+            echo "Image not able to be built" >> build_result.txt
+          fi
+
+      - name: Upload image artifact
+        uses: actions/upload-artifact@v3
+        with:
+          name: meals-api
+          path: /tmp/image.tar
+
+      - name: Upload result file
+        if: always()
+        uses: actions/upload-artifact@v3
+        with:
+          name: meals-api-build-result.txt
+          path: build_result.txt
+  test:
+    needs: build
+    runs-on: ubuntu-latest
+    steps:
+    #1. download image and run it in a container
+    - name: Download artifact
+      uses: actions/download-artifact@v3
+    - name: Load Docker images from previous workflows
+      run: docker load --input meals-api/image.tar
+    - name: Check meals-api image is loaded
+      run: docker image ls
+    - name: Run image in container
+      run: docker run -p8000:8000 meals-api &
+    - name: Check container is running

```

Comment:
I think we need to download the artifact and re-upload it with the appropriate new lines to it (container running and test results)
we need to add to log.txt these two things:


- If the container is successfully started then “Container up and
running”. Otherwise (the container not started successfully) “Container failed
to run”.
- If the tests are successful then “tests succeeded”. Otherwise (at least
one test failed) then “tests failed”.

<img width="627" alt="Screenshot 2023-06-23 at 1 54 12 PM" src="https://github.com/DanielItkes/Meals-API/assets/54701359/38967ec3-8aff-4378-b78d-73ae5a9853f1">

Reply:
just need to keep this in mind:


> If the workflow terminates because of an error
> in a step, then only the previous logs and the
> log relating to the error step (if there is one)
> needs to be included. For example, if the
> image is not successfully built, then lines 1-3
> would be in the log but not line 4. Similarly,
> if the container fails to start, then lines 1-4
> would be in the log, but not line 5.